// expo-image-cropper/src/ControlBar.jsx
import { View, StyleSheet } from "react-native";
import { IconButton } from "./components/IconButton";
import { useContext } from "react";
import { EditorContext } from "./context/editor";
import { usePerformCrop } from "./customHooks/usePerformCrop";
import { useRecoilState } from "recoil";
import { isEditState } from "./Store";

function ControlBar() {
  const [isEdit, setIsEdit] = useRecoilState(isEditState);
  const { onBackPress, onSave } = useContext(EditorContext);
  const performCrop = usePerformCrop();

  const onEditDone = async () => {
    await performCrop();
    setIsEdit(true);
  };

  return (
    <View style={styles.container}>
      <IconButton
        iconID={!isEdit ? "x" : "arrow-back"}
        text={!isEdit ? "Cancel" : "Back"}
        onPress={() => {
          onBackPress();
          setIsEdit(false);
        }}
      />
      {!isEdit ? (
        <IconButton iconID="crop" text="Crop" onPress={onEditDone} />
      ) : (
        <IconButton iconID="done" text="Save" onPress={onSave} />
      )}
    </View>
  );
}

export { ControlBar };

const styles = StyleSheet.create({
  container: {
    width: "100%",
    height: 80,
    backgroundColor: "#333",
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    paddingHorizontal: 4,
  },
});
